src/Plugin/AddPathPlugin.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Http\Client\Common\Plugin;
+
+use Http\Client\Common\Plugin;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Add base path to the request URI. Useful for base API URLs like http://domain.com/api.
+ *
+ * @author Sullivan Senechal <soullivaneuh@gmail.com>
+ */
+final class AddPathPlugin implements Plugin
+{
+    /**
+     * @var UriInterface
+     */
+    private $host;
+
+    /**
+     * @param UriInterface $host
+     */
+    public function __construct(UriInterface $host)
+    {
+        if ($host->getPath() === '') {
+            throw new \LogicException('Host path can not be empty');
+        }
+
+        $this->host = $host;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        $request = $request->withUri($request->getUri()
+            ->withPath($this->host->getPath().$request->getUri()->getPath())
```

Comment:
what about anything ending on /? notably http://domain.com/ but also http://domain.com/path/ - both would end up in double leading slash, which is very likely not what the user wanted.

this is a rather annoying edge case. i think we should somehow handle it (maybe validate and InvalidArgumentException) to avoid wtf moments (some servers will handle the example.com//something paths, others wont). if somebody actually wants to create double slashes, they can write their own plugin, its a rare edgecase.